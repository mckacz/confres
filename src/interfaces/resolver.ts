/**
 * Loads and process the configuration.
 * Decorates eventual errors with references to the causing value.
 */
export interface Resolver<T = unknown> {
  /**
   * Loads and processes the configuration.
   * Returns final configuration object.
   */
  resolve(): T
}
