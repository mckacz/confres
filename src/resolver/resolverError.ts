import type { Loader} from '../interfaces/loader'
import type { Processor } from '../interfaces/processor'
import type{ Reference } from '../interfaces/reference'
import { ConfresError } from '../common/confresError'
import { ProcessorError, ProcessorErrorType } from '../processors/processorError'

/**
 * Resolver error.
 */
export class ResolverError extends ConfresError {
  /**
   * Create a new instance of resolver error.
   *
   * @param err Original error.
   * @param reporter Reporter of the error - either a loader or processor instance.
   * @param path Object path of value that caused the error (if available).
   * @param references Reference to value that caused the error (for invalid value error)
   *                   or list of references where value can be defined (for undefined value error).
   */
  constructor(
    err: Error | unknown,
    readonly reporter?: Loader<unknown> | Processor<unknown, unknown>,
    readonly path?: string[],
    readonly references: Reference[] = [],
  ) {
    super(
      err instanceof Error ? err.message : `Configuration resolve error: ${String(err)}`,
      err,
    )
  }

  /**
   * True if it holds undefined value error.
   */
  get isUndefinedError() {
    return this.error instanceof ProcessorError
      && this.error.type === ProcessorErrorType.undefinedValue
  }
}
