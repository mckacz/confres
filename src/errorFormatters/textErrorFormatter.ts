import _ from 'lodash'
import { Reference } from '../loaders/loader.js'
import { ResolverError } from '../resolver/resolverError.js'
import { ErrorFormatter } from './errorFormatter.js'
import type { DeepPartial } from '../helpers.js'

export interface TextErrorFormatterTemplates {
  header: string
  causeDescription: string
  hintListHeader: string
  hintListItem: string
  hintFull: string
  hintContainerOnly: string
  lineSeparator: string
}

export interface TextErrorFormatterOptions {
  templates: TextErrorFormatterTemplates
}

export enum TextErrorFormatterPlaceholders {
  message = 'message',
  hint = 'hint',
  container = 'container',
  identifier = 'identifier',
}

const defaultTemplates: TextErrorFormatterTemplates = {
  header:            'Configuration error: [message]',
  causeDescription:  'The causing value is defined in [hint]',
  hintListHeader:    'The value can be defined in:',
  hintListItem:      '  - [hint]',
  hintFull:          '[container] as [identifier]',
  hintContainerOnly: '[container]',
  lineSeparator:     '\n',
}

const defaultOptions: TextErrorFormatterOptions = {
  templates: defaultTemplates,
}

export class TextErrorFormatter implements ErrorFormatter {
  private readonly templates: TextErrorFormatterTemplates

  constructor(options: DeepPartial<TextErrorFormatterOptions> = {}) {
    const { templates }: TextErrorFormatterOptions = _.merge({}, defaultOptions, options)

    this.templates = templates
  }

  format(err: ResolverError | unknown): string | undefined {
    if (!(err instanceof ResolverError)) {
      return
    }

    const lines: (string | undefined)[] = [
      this.renderHeader(err),
    ]

    if (err.isUndefinedError) {
      lines.push(...this.renderHintList(err))
    } else {
      lines.push(this.renderCause(err))
    }

    return lines
      .filter(line => line !== undefined)
      .join(this.templates.lineSeparator)
  }

  protected renderHeader(err: ResolverError): string {
    return this.renderTemplate(this.templates.header, {
      [TextErrorFormatterPlaceholders.message]: err.message,
    })
  }

  protected renderCause(err: ResolverError): string | undefined {
    if (!err.references[0]) {
      return
    }

    return this.renderTemplate(this.templates.causeDescription, {
      [TextErrorFormatterPlaceholders.hint]: this.renderHint(err.references[0]),
    })
  }

  protected renderHintList(err: ResolverError): string[] {
    const lines: string[] = []

    lines.push(this.renderTemplate(this.templates.hintListHeader, {}))

    for (const ref of err.references) {
      lines.push(this.renderTemplate(this.templates.hintListItem, {
        [TextErrorFormatterPlaceholders.hint]: this.renderHint(ref),
      }))
    }

    return lines
  }

  protected renderHint(ref: Reference): string {
    if (ref.identifier) {
      return this.renderTemplate(this.templates.hintFull, {
        [TextErrorFormatterPlaceholders.container]:  ref.container,
        [TextErrorFormatterPlaceholders.identifier]: ref.identifier,
      })
    } else {
      return this.renderTemplate(this.templates.hintContainerOnly, {
        [TextErrorFormatterPlaceholders.container]: ref.container,
      })
    }
  }

  protected renderTemplate(template: string, dictionary: Record<string, string>): string {
    for (const [placeholder, value] of Object.entries(dictionary)) {
      template = template.replace(`[${placeholder}]`, value)
    }

    return template
  }
}
