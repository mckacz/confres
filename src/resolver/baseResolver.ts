import _ from 'lodash'
import type { Loader} from '../interfaces/loader'
import type { Processor } from '../interfaces/processor'
import { ProcessorError, ProcessorErrorType } from '../processors/processorError'
import { ResolverError } from './resolverError'
import { Resolver } from '../interfaces/resolver'
import type { Reference } from '../interfaces/reference'

/**
 * Common parts of AsyncResolver and SyncResolver.
 */
export abstract class BaseResolver<Type, ReturnType = Type> implements Resolver<ReturnType> {
  /**
   * Create a new instance of resolver.
   *
   * @param loaders List of loaders to use. Configuration acquired from the loaders
   *                will be merged in order from left to right.
   *
   * @param processors List of processors to be applied on loaded configuration object.
   */
  constructor(
    protected readonly loaders: Loader<unknown | undefined>[],
    protected readonly processors: Processor<unknown, Type>[] = [],
  ) {
  }

  /**
   * Loads and processes the configuration.
   * Returns final configuration object.
   */
  abstract resolve(): ReturnType

  /**
   * Decorates processing error with references to the causing value if available.
   *
   * @param err Error object.
   * @param sources Source configuration objects.
   * @param processor Processor instance that reported the error.
   */
  protected decorateError(err: ProcessorError | unknown, sources: unknown[], processor: Processor<unknown, unknown>) {
    if (!(err instanceof ProcessorError) || !err.path) {
      return new ResolverError(err, processor)
    }

    if (err.type === ProcessorErrorType.undefinedValue) {
      return new ResolverError(
        err,
        processor,
        undefined,
        err.path,
        this.resolveReferencesForUndefinedValue(err.path),
      )
    }

    const loader = this.findLoaderOfPath(err.path, sources)

    return new ResolverError(
      err,
      processor,
      loader,
      err.path,
      this.removeEmptyReferences([loader?.referenceFor(err.path)]),
    )
  }

  /**
   * Tries to find a loader that has loaded a value for given object path.
   *
   * @param path Object path.
   * @param sources Source configuration objects.
   */
  private findLoaderOfPath(path: string, sources: unknown[]): Loader<unknown> | undefined {
    const sourceIndex = _.findLastIndex(sources, sourceValue => _.get(sourceValue, path) !== undefined)

    if (sourceIndex !== -1) {
      return this.loaders[sourceIndex]
    }
  }

  /**
   * Resolves a list of references where the undefined value can be defined.
   *
   * @param path Object path.
   */
  private resolveReferencesForUndefinedValue(path: string): Reference[] {
    return this.removeEmptyReferences(
      this.loaders.map(
        loader => loader.referenceFor(path),
      ),
    )
  }

  /**
   * Removes empty references from the list.
   *
   * @param references List of references.
   */
  private removeEmptyReferences(references: Array<Reference | undefined>): Reference[] {
    return references.filter(ref => ref !== undefined) as Reference[]
  }
}
