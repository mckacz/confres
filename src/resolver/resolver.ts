import _ from 'lodash'
import type { Loader, Reference } from '../loaders/loader'
import type { Processor } from '../processors/processor'
import { ResolverError } from './resolverError'
import { ProcessorError, ProcessorErrorType } from '../processors/processorError'

/**
 * Loads and process the configuration.
 * Decorates eventual errors with references to the causing value.
 */
export class Resolver<T = unknown> {
  /**
   * Create a new instance of resolver.
   *
   * @param loaders List of loaders to use. Configuration acquired from the loaders
   *                will be merged in order from left to right.
   *
   * @param processors List of processors to be applied on loaded configuration object.
   */
  constructor(
    private readonly loaders: Loader<unknown | undefined>[],
    private readonly processors: Processor<unknown, T>[] = [],
  ) {
  }

  /**
   * Loads and processes the configuration.
   * Returns final configuration object.
   */
  async resolve(): Promise<T> {
    const sources = await this.load()
    const resolved = await this.process(sources)

    return resolved
  }

  /**
   * Loads configurations using loaders.
   */
  private async load(): Promise<unknown[]> {
    const sources: unknown[] = []

    for (const loader of this.loaders) {
      try {
        sources.push(await loader.load())
      } catch (ex) {
        throw new ResolverError(ex, loader)
      }
    }
    return sources
  }

  /**
   * Merges and processes the configurations using processors.
   *
   * @param sources Source configuration objects.
   */
  private async process(sources: unknown[]): Promise<T> {
    let merged = _.merge({}, ...sources) as T

    for (const processor of this.processors) {
      try {
        merged = await processor.process(merged)
      } catch (ex) {
        throw this.decorateError(ex, sources, processor)
      }
    }

    return merged
  }

  /**
   * Decorates processing error with references to the causing value if available.
   *
   * @param err Error object.
   * @param sources Source configuration objects.
   * @param processor Processor instance that reported the error.
   */
  private decorateError(err: ProcessorError | unknown, sources: unknown[], processor: Processor<unknown, unknown>) {
    if (!(err instanceof ProcessorError) || !err.path) {
      return new ResolverError(err, processor)
    }

    if (err.type === ProcessorErrorType.undefinedValue) {
      return new ResolverError(
        err,
        processor,
        undefined,
        err.path,
        this.resolveReferencesForUndefinedValue(err.path)
      )
    }

    const loader = this.findLoaderOfPath(err.path, sources)

    return new ResolverError(
      err,
      processor,
      loader,
      err.path,
      this.removeEmptyReferences([loader?.referenceFor(err.path)])
    )
  }

  /**
   * Tries to find a loader that has loaded a value for given object path.
   *
   * @param path Object path.
   * @param sources Source configuration objects.
   */
  private findLoaderOfPath(path: string, sources: unknown[]): Loader<unknown> | undefined {
    const sourceIndex = _.findLastIndex(sources, sourceValue => _.get(sourceValue, path) !== undefined)

    if (sourceIndex !== -1) {
      return this.loaders[sourceIndex]
    }
  }

  /**
   * Resolves a list of references where the undefined value can be defined.
   *
   * @param path Object path.
   */
  private resolveReferencesForUndefinedValue(path: string): Reference[] {
    return this.removeEmptyReferences(
      this.loaders.map(
        loader => loader.referenceFor(path)
      )
    )
  }

  /**
   * Removes empty references from the list.
   *
   * @param references List of references.
   */
  private removeEmptyReferences(references: Array<Reference | undefined>): Reference[] {
    return references.filter(ref => ref !== undefined)
  }
}
