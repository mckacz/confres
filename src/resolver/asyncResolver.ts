import _ from 'lodash'
import { ResolverError } from './resolverError'
import { BaseResolver } from './baseResolver'

/**
 * Asynchronous resolver.
 */
export class AsyncResolver<T = unknown> extends BaseResolver<T, Promise<T>> {
  /**
   * Loads and processes the configuration.
   * Returns final configuration object.
   */
  async resolve(): Promise<T> {
    const sources = await this.load()
    const resolved = await this.process(sources)

    return resolved
  }

  /**
   * Loads configurations using loaders.
   */
  private async load(): Promise<unknown[]> {
    const sources: unknown[] = []

    for (const loader of this.loaders) {
      try {
        sources.push(await loader.load())
      } catch (ex) {
        throw new ResolverError(ex, loader)
      }
    }
    return sources
  }

  /**
   * Merges and processes the configurations using processors.
   *
   * @param sources Source configuration objects.
   */
  private async process(sources: unknown[]): Promise<T> {
    let merged = _.merge({}, ...sources) as T

    for (const processor of this.processors) {
      try {
        merged = await processor.process(merged)
      } catch (ex) {
        throw this.decorateError(ex, sources, processor)
      }
    }

    return merged
  }
}
