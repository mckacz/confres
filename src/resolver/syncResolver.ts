import { ResolverError } from './resolverError'
import { BaseResolver } from './baseResolver'
import { MergedDataContainer } from '../dataContainers/mergedDataContainer'
import type { DataContainer } from '../interfaces/dataContainer'

/**
 * Synchronous resolver.
 */
export class SyncResolver<T = unknown> extends BaseResolver<T> {
  /**
   * Loads and processes the configuration synchronously.
   * Returns final configuration object.
   */
  resolve(): T {
    const sources = this.load()
    const resolved = this.process(sources)

    return resolved
  }

  /**
   * Loads configurations using loaders.
   */
  private load(): DataContainer<unknown> {
    const sources: DataContainer<unknown>[] = []

    for (const loader of this.loaders) {
      try {
        if (this.assertSync(loader, 'loadSync')) {
          sources.push(loader.loadSync())
        }
      } catch (ex) {
        throw new ResolverError(ex, loader)
      }
    }

    return new MergedDataContainer(sources)
  }

  /**
   * Merges and processes the configurations using processors.
   *
   * @param dataContainer Data container.
   */
  private process(dataContainer: DataContainer<unknown>): T {
    let merged = dataContainer.value as T

    for (const processor of this.processors) {
      try {
        if (this.assertSync(processor, 'processSync')) {
          merged = processor.processSync(merged)
        }
      } catch (ex) {
        throw this.decorateError(ex, dataContainer, processor)
      }
    }

    return merged
  }

  /**
   * Asserts that a given method is implemented in the object.
   *
   * @param thing Object to check
   * @param method Method name to check
   */
  private assertSync<T extends object, M extends keyof T>(thing: T, method: M): thing is T & Required<Pick<T, M>> {
    if (typeof thing[method] !== 'function') {
      throw new TypeError(`'${thing.constructor.name}' does not support synchronous mode`)
    }

    return true
  }
}
