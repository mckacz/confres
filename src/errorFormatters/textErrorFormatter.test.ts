import { ResolverError } from '../resolver/resolverError'
import { TextErrorFormatter, TextErrorFormatterTemplates } from './textErrorFormatter'
import {
  processingErrorMock,
  processingErrorWithoutReferencesMock,
  undefinedErrorMock,
  undefinedErrorWithoutReferencesMock,
} from '../../fixtures/errors'

describe('TextErrorFormatter', () => {
  describe('Default templates', () => {
    test.each([
      ['Processing error', processingErrorMock],
      ['Processing error without references', processingErrorWithoutReferencesMock],
      ['Undefined value error', undefinedErrorMock],
      ['Undefined value error without references', undefinedErrorWithoutReferencesMock],
    ])('%s', (_: string, error: ResolverError) => {
      const formatter = new TextErrorFormatter()
      expect(formatter.format(error)).toMatchSnapshot()
    })
  })

  describe('Custom templates', () => {
    const customTemplates: TextErrorFormatterTemplates = {
      header:            'Oh no! [message]',
      causeDescription:  'Update the value of [hint]',
      hintListHeader:    'You can provide the value using:',
      hintListItem:      '  * [hint]',
      hintFull:          '[identifier] ([container])',
      hintContainerOnly: '[container]',
      lineSeparator:     '\n',
    }

    test.each([
      ['Processing error', processingErrorMock],
      ['Processing error without references', processingErrorWithoutReferencesMock],
      ['Undefined value error', undefinedErrorMock],
      ['Undefined value error without references', undefinedErrorWithoutReferencesMock],
    ])('%s', (_: string, error: ResolverError) => {
      const formatter = new TextErrorFormatter(customTemplates)
      expect(formatter.format(error)).toMatchSnapshot()
    })
  })

  test('ignore non-ResolverError', () => {
    const formatter = new TextErrorFormatter()

    expect(formatter.format(new Error('Oh no!'))).toBeUndefined()
  })
})
