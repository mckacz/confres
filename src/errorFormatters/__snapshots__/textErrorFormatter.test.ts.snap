// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextErrorFormatter Custom templates Processing error 1`] = `
"Oh no! The value \`11\` must be grater than or equal to \`42\`
Update the value of --obj.foo (Command line arguments)"
`;

exports[`TextErrorFormatter Custom templates Processing error without references 1`] = `"Oh no! The value \`11\` must be grater than or equal to \`42\`"`;

exports[`TextErrorFormatter Custom templates Undefined value error 1`] = `
"Oh no! The value is required
You can provide the value using:
  * OBJ__FOO (Environment variables)
  * --obj.foo (Command line arguments)"
`;

exports[`TextErrorFormatter Custom templates Undefined value error without references 1`] = `"Oh no! The value is required"`;

exports[`TextErrorFormatter Default templates Processing error 1`] = `
"Configuration error: The value \`11\` must be grater than or equal to \`42\`
The causing value is defined in Command line arguments as --obj.foo"
`;

exports[`TextErrorFormatter Default templates Processing error without references 1`] = `"Configuration error: The value \`11\` must be grater than or equal to \`42\`"`;

exports[`TextErrorFormatter Default templates Undefined value error 1`] = `
"Configuration error: The value is required
The value can be defined in:
  - Environment variables as OBJ__FOO
  - Command line arguments as --obj.foo"
`;

exports[`TextErrorFormatter Default templates Undefined value error without references 1`] = `"Configuration error: The value is required"`;
